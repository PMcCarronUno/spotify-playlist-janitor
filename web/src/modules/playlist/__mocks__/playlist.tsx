import React from 'react';
export const Playlist = ({ id }: { id: string }) => <div>Playlist: {id} Component</div>;
