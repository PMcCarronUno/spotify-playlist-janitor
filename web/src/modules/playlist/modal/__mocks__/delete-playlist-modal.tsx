import React from 'react';
export const DeletePlaylistModalView = () => <div>Delete Playlist Modal View</div>