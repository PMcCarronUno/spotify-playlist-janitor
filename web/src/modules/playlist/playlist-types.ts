import { Playlist } from "shared/types";

export type DetailedPlaylist = Playlist & {
  skipThreshold?: number;
  ignoreInitialSkips: boolean;
  autoCleanupLimit?: number;
};
