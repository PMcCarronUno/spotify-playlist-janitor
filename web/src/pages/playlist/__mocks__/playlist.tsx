import React from 'react';
export const Playlist = () => <div>Playlist</div>;