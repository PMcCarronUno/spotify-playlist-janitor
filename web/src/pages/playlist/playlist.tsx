import { useParams } from "react-router-dom";
import styled from "styled-components";
import { Playlist as PlaylistInitialise } from "modules/playlist";
import { HEADER_HEIGHT, HEADER_PADDING } from "shared/constants";

export const Playlist = () => {
  let { id } = useParams();
  return (
    <Content>
      <PlaylistInitialise {...{ id }} />
    </Content>
  );
};

const Content = styled.div`
  display: flex;
  justify-content: center;
  height: calc(
    100vh - ${HEADER_HEIGHT} - ${HEADER_PADDING} - ${HEADER_PADDING}
  );
`;

Playlist.displayName = "Playlist";
