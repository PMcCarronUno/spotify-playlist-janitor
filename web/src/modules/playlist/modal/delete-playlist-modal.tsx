import styled from "styled-components";
import { RED } from "shared/constants";
import { Text, SecondaryText, SubTitle } from "shared/components/typography";
import { PrimaryButton, SecondaryButton } from "shared/components/button";
import { DetailedPlaylist } from "../playlist-types";

type DeletePlaylistModalViewProps = {
  onSubmit(event: React.FormEvent<HTMLFormElement>): Promise<void>;
  playlist: DetailedPlaylist | undefined;
  modalSaving: boolean;
  modalError?: string;
  onModalClose(): void;
};

export const DeletePlaylistModalView = ({
  onSubmit,
  playlist,
  modalSaving,
  modalError,
  onModalClose,
}: DeletePlaylistModalViewProps) => {
  return (
    <form {...{ onSubmit }} autoComplete="off">
      <ModalContainer>
        <SubTitle style={{ marginBottom: 8 }}>
          Remove playlist
        </SubTitle>
        <Text style={{ marginBottom: 8 }}>
          Are you sure you want to remove {playlist?.name} from monitoring?
          This will remove all skiiped track records for this playlist.
        </Text>
        <div className="rows">
          {modalError && (
            <div
              {...{
                className: "row error",
                id: "modal-error",
                "data-testid": "modal-error",
              }}
            >
              <SecondaryText>{modalError}</SecondaryText>
            </div>
          )}
        </div>
        <div
          style={{
            display: "flex",
            flexDirection: "row",
            justifyContent: "flex-end",
          }}
        >
          <SecondaryButton
            {...{
              className: "secondary",
              style: { margin: "0px 4px" },
              type: "button",
              id: "cancel-modal-button",
              "data-testid": "cancel-modal-button",
              onClick: onModalClose,
              disabled: modalSaving,
            }}
          >
            Cancel
          </SecondaryButton>
          <PrimaryButton
            {...{
              className: "primary",
              style: { margin: "0px 4px" },
              type: "submit",
              id: "confirm-modal-button",
              "data-testid": "confirm-modal-button",
              disabled: modalSaving,
            }}
          >
            Confirm
          </PrimaryButton>
        </div>
      </ModalContainer>
    </form>
  );
};

const ModalContainer = styled.div`
  width: 360px;

  .rows {
    margin-bottom: 8px;
  }

  .row {
    display: flex;
    flex-direction: row;
    align-items: center;
    margin-bottom: 8px;
  }

  .number-input {
    width: 55px !important;
  }

  .toggle-input {
    margin-right: 10px;
  }

  .auto-delete-post-text {
    margin-right: 10px;
  }

  .error {
    color: ${RED};
  }
`;

DeletePlaylistModalView.displayName = "DeletePlaylistModalView";
