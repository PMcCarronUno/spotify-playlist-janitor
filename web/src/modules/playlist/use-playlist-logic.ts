import { useCallback, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";

import { useDataApi } from "api/data-api";
import { useSpotifyApi } from "api/spotify-api";

import { DetailedPlaylist } from "./playlist-types";
import { useModal } from "shared/components/modal";
import { DEFAULT_IMAGE, HOME } from "shared/constants";

type PlaylistProps = {
  id: string;
};

export const usePlaylistLogic = ({ id }: PlaylistProps) => {
  const {
    getDatabasePlaylist,
    updateDatabasePlaylist,
    deleteDatabasePlaylist,
  } = useDataApi();
  const { getSpotifyPlaylist } = useSpotifyApi();
  const navigate = useNavigate();

  const [notFound, setNotFound] = useState(false);
  const [loading, setLoading] = useState(false);
  const [playlist, setPlaylist] = useState<DetailedPlaylist | undefined>(
    undefined
  );

  const [editSaving, setEditSaving] = useState(false);
  const [editError, setEditError] = useState<string | undefined>(undefined);

  const [deleting, setDeleting] = useState(false);
  const [deleteError, setDeleteError] = useState<string | undefined>(undefined);

  const {
    isOpen: editOpen,
    onOpen: onEditOpen,
    onClose: onEditModalClose,
  } = useModal();
  const {
    isOpen: deleteOpen,
    onOpen: onDeleteOpen,
    onClose: onDeleteClose,
  } = useModal();

  const onEditClose = useCallback(() => {
    setEditError(undefined);
    onEditModalClose();
  }, [setEditError, onEditModalClose]);

  const getPlaylistInfo = useCallback(
    async (id: string) => {
      try {
        setLoading(true);
        setPlaylist(undefined);
        const databasePlaylistsResponse = await getDatabasePlaylist(id);

        const spotifyPlaylistsResponse = await getSpotifyPlaylist(id);

        setPlaylist({
          id: databasePlaylistsResponse.data.id,
          skipThreshold: databasePlaylistsResponse.data.skipThreshold,
          ignoreInitialSkips: databasePlaylistsResponse.data.ignoreInitialSkips,
          autoCleanupLimit: databasePlaylistsResponse.data.autoCleanupLimit,
          name: spotifyPlaylistsResponse.data.name,
          href: spotifyPlaylistsResponse.data.href,
          image: spotifyPlaylistsResponse.data.images[0] ?? DEFAULT_IMAGE,
        });

        setLoading(false);
      } catch (e: any) {
        console.log("error", e)
        if (e?.response?.status === 404) {
          setNotFound(true);
        }
        setLoading(false);
      }
    },
    [setLoading, getDatabasePlaylist, getSpotifyPlaylist, setNotFound]
  );

  useEffect(() => {
    if (!!id) {
      getPlaylistInfo(id);
    }
  }, [getPlaylistInfo, id]);

  const onEditSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      setEditError(undefined);
      const skipThreshold = (event.currentTarget[0] as HTMLInputElement).value;
      const ignoreInitialSkips = (event.currentTarget[1] as HTMLInputElement)
        .checked;
      const autoDeleteTracksAfter = (event.currentTarget[2] as HTMLInputElement)
        .value;

      if (!!autoDeleteTracksAfter && parseInt(autoDeleteTracksAfter) === 0) {
        setEditError("Auto-delete must be empty or greater than 0");
        return;
      }

      try {
        setEditSaving(true);

        const request = {
          skipThreshold: !!skipThreshold ? parseInt(skipThreshold) : undefined,
          ignoreInitialSkips: ignoreInitialSkips,
          autoCleanupLimit: !!autoDeleteTracksAfter
            ? parseInt(autoDeleteTracksAfter)
            : undefined,
        };

        await updateDatabasePlaylist(id, request);
        setEditSaving(false);
        onEditClose();
        await getPlaylistInfo(id);
      } catch (e: any) {
        setEditError(e.response?.data?.message || "Unknown error");
        setEditSaving(false);
      }
    },
    [id, updateDatabasePlaylist, getPlaylistInfo, onEditClose]
  );

  const onDeleteSubmit = useCallback(
    async (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();

      try {
        setDeleting(true);
        await deleteDatabasePlaylist(id);
        setDeleting(false);
        onDeleteClose();
        navigate(HOME)
      } catch (e: any) {
        setDeleteError(e.response?.data?.message || "Unknown error");
        setDeleting(false);
      }
    },
    [id, deleteDatabasePlaylist, navigate, onDeleteClose]
  );

  return {
    loading,
    notFound,
    playlist,
    editOpen,
    onEditOpen,
    onEditClose,
    onEditSubmit,
    editError,
    editSaving,
    deleting,
    deleteOpen, 
    onDeleteOpen, 
    onDeleteClose,
    onDeleteSubmit,
    deleteError,
  } as const;
};
