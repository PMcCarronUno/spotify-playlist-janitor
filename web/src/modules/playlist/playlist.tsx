import styled from "styled-components";
import {
    SKELETON_PLAYLIST_BASE,
    SKELETON_HIGHLIGHT,
    WHITE,
} from "shared/constants";
import { usePlaylistLogic } from "./use-playlist-logic";
import { Menu, MenuItem } from "shared/components/menu";
import { VscEdit, VscEllipsis, VscLinkExternal, VscTrash } from "react-icons/vsc";
import { SecondaryText, Text, Title } from "shared/components/typography";
import { Skeleton, SkeletonTheme } from "shared/components/skeleton";
import { Modal } from "shared/components/modal";
import { EditPlaylistModalView } from "./modal";
import { DeletePlaylistModalView } from "./modal/delete-playlist-modal";
import { PlaylistTabs } from "./tabs/playlist-tabs";


type PlaylistProps = {
    id: string;
};

const skeleton = <SkeletonTheme
    baseColor={SKELETON_PLAYLIST_BASE}
    highlightColor={SKELETON_HIGHLIGHT}
    height="100%"
>
    <Skeleton />
</SkeletonTheme>;

export const Playlist = ({ id }: PlaylistProps) => {
    const {
        loading,
        notFound,
        playlist,
        editOpen,
        onEditOpen,
        onEditClose,
        onEditSubmit,
        editError,
        editSaving,
        deleting,
        deleteOpen,
        onDeleteOpen,
        onDeleteClose,
        onDeleteSubmit,
        deleteError,
    } = usePlaylistLogic({ id });

    if (notFound) {
        return <PageContainer>
            <Text {...{
                id: "not-found",
                "data-testid": "not-found",
            }}>
                Not Found
            </Text>
        </PageContainer>;
    }

    return (
        <PageContainer>
            <div className="header">
                <div className="playlist">
                    <div className="playlist-image" id="playlist-image">
                        {loading ?
                            skeleton
                            : (
                                <img
                                    width={240}
                                    height={240}
                                    src={playlist?.image?.url}
                                    alt={playlist?.name}
                                />
                            )}
                    </div>
                    <div className="playlist-details">
                        <Title className={`playlist-title${loading ? "-loading" : ""}`}>
                            {!!playlist
                                ?
                                <>
                                    {playlist?.name}
                                    <a className="icon-link"
                                        href={playlist?.href}
                                        target="_blank"
                                        rel="noreferrer">
                                        <VscLinkExternal style={{ cursor: "pointer", fontSize: 28 }} />
                                    </a>
                                </>
                                :
                                skeleton
                            }
                        </Title>
                        <div
                            className={`playlist-details-item${loading ? "-loading" : ""}`}
                        >
                            {playlist?.skipThreshold ? (
                                <>
                                    <SecondaryText>Skip Threshold:&nbsp;</SecondaryText>
                                    <Text>{playlist?.skipThreshold} s</Text>
                                </>
                            ) : skeleton}
                        </div>

                        <div
                            className={`playlist-details-item${loading ? "-loading" : ""}`}
                        >
                            {playlist?.skipThreshold ? (
                                <>
                                    <SecondaryText>Ignore initial skips:&nbsp;</SecondaryText>
                                    <Text>{playlist?.ignoreInitialSkips ? "True" : "False"}</Text>
                                </>
                            ) : skeleton}
                        </div>

                        <div
                            className={`playlist-details-item${loading ? "-loading" : ""}`}
                        >
                            {playlist ? (
                                <>
                                    <SecondaryText>Auto-delete after:&nbsp;</SecondaryText>
                                    {playlist?.autoCleanupLimit ? (
                                        <>
                                            <Text>{playlist?.autoCleanupLimit}&nbsp;</Text>
                                            <SecondaryText>skips</SecondaryText>
                                        </>
                                    ) : (
                                        <Text>No auto-delete</Text>
                                    )}
                                </>
                            ) : skeleton}
                        </div>
                    </div>
                </div>
                <div className="playlist-menu">
                    <Menu
                        menuButton={
                            <div>
                                <VscEllipsis style={{ cursor: "pointer", fontSize: 28 }} />
                            </div>
                        }
                    >
                        <MenuItem {...{ onClick: onEditOpen }}>
                            <VscEdit style={{ cursor: "pointer", marginRight: 8 }} />
                            Edit
                        </MenuItem>
                        <MenuItem {...{ onClick: onDeleteOpen }}>
                            <VscTrash style={{ cursor: "pointer", marginRight: 8 }} />
                            Delete
                        </MenuItem>
                    </Menu>
                </div>
            </div>

            <div className="content">
                <PlaylistTabs {...{ id, loading }} />
            </div>

            <Modal
                {...{
                    isOpen: editOpen,
                    onClose: onEditClose,
                    label: "Edit Playlist Modal",
                }}
            >
                <EditPlaylistModalView
                    {...{
                        onSubmit: onEditSubmit,
                        playlist,
                        modalSaving: editSaving,
                        modalError: editError,
                        onModalClose: onEditClose,
                    }}
                />
            </Modal>
            <Modal
                {...{
                    isOpen: deleteOpen,
                    onClose: onDeleteClose,
                    label: "Delete Playlist Modal",
                }}
            >
                <DeletePlaylistModalView
                    {...{
                        onSubmit: onDeleteSubmit,
                        playlist,
                        modalSaving: deleting,
                        modalError: deleteError,
                        onModalClose: onDeleteClose,
                    }}
                />
            </Modal>
        </PageContainer>
    );
};

const PageContainer = styled.div`
  display: flex;
  flex-direction: column;
  color: ${WHITE};
  align-items: center;

  width: 100%;
  max-width: 1825px;
  padding-top: 16px;

  .header {
    display: flex;
    flex-direction: row;
    justify-content: space-around;
    width: 100%;

    margin-bottom: 24px;
  }

  .playlist {
    display: flex;
    flex-direction: row;
  }

  .playlist-image {
    width: 240px;
    height: 240px;
  }

  .playlist-details {
    padding-left: 32px;

    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;

    width: 400px;
  }

  .playlist-title {
    margin-bottom: 24px;
  }

  .playlist-title-loading {
    height: 42px;
    width: 360px;
    margin-bottom: 24px;
  }

  .playlist-details-item {
    margin-bottom: 12px;

    display: flex;
  }

  .playlist-details-item-loading {
    height: 22px;
    width: 180px;
    margin-bottom: 12px;
  }

  .content {
    width: 100%;
  }

  .icon-link{
    color: ${WHITE};
    margin-left: 12px;

    svg {
        margin-bottom: -3px;
    }
  }
`;

Playlist.displayName = "Playlist";
