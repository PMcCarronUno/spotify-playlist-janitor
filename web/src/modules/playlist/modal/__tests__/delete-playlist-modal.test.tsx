import { fireEvent, getByTestId, render } from "@testing-library/react";

import { DeletePlaylistModalView } from "../delete-playlist-modal";
import { playlist } from "shared/mock-data/playlist";

describe("<DeletePlaylistModalView />", () => {
  let container: HTMLElement;
  let props: any = {
    onSubmit: jest.fn(),
    playlist,
    modalSaving: false,
    modalError: undefined,
    onModalClose: jest.fn(),
  };

  beforeEach(async () => {
    ({ container } = render(<DeletePlaylistModalView {...props} />));
  });

  it("should render DeletePlaylistModalView component", () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render DeletePlaylistModalView component with an error", () => {
    props = {
      ...props,
      modalError: "MODAL_ERROR",
    };

    ({ container } = render(<DeletePlaylistModalView {...props} />));

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should call onSubmit function when Confirm button clicked", () => {
    const confirm = getByTestId(container, "confirm-modal-button");
    fireEvent.click(confirm);

    expect(props.onSubmit).toHaveBeenCalled();
  });

  it("should call onModalClose function when Cancel button clicked", () => {
    const cancel = getByTestId(container, "cancel-modal-button");
    fireEvent.click(cancel);

    expect(props.onModalClose).toHaveBeenCalled();
  });
});
