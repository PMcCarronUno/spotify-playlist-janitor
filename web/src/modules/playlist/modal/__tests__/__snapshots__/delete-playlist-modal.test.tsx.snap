// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DeletePlaylistModalView /> should render DeletePlaylistModalView component 1`] = `
<form
  autocomplete="off"
>
  <div
    class="sc-fsYfxw jIlICe"
  >
    <div
      class="sc-guDMob jLqGJD"
      style="margin-bottom: 8px;"
    >
      Remove playlist
    </div>
    <div
      class="sc-dmyDmy dcXriK"
      style="margin-bottom: 8px;"
    >
      Are you sure you want to remove 
      Playlist 1
       from monitoring?
    </div>
    <div
      class="rows"
    />
    <div
      style="display: flex; flex-direction: row; justify-content: flex-end;"
    >
      <button
        class="sc-jTQDJr sc-iBdoyZ gmttQv bbczBv secondary"
        data-testid="cancel-modal-button"
        id="cancel-modal-button"
        style="margin: 0px 4px;"
        type="button"
      >
        Cancel
      </button>
      <button
        class="sc-jTQDJr sc-gLLuHO gmttQv lckOnr primary"
        data-testid="confirm-modal-button"
        id="confirm-modal-button"
        style="margin: 0px 4px;"
        type="submit"
      >
        Confirm
      </button>
    </div>
  </div>
</form>
`;

exports[`<DeletePlaylistModalView /> should render DeletePlaylistModalView component with an error 1`] = `
<form
  autocomplete="off"
>
  <div
    class="sc-fsYfxw jIlICe"
  >
    <div
      class="sc-guDMob jLqGJD"
      style="margin-bottom: 8px;"
    >
      Remove playlist
    </div>
    <div
      class="sc-dmyDmy dcXriK"
      style="margin-bottom: 8px;"
    >
      Are you sure you want to remove 
      Playlist 1
       from monitoring?
    </div>
    <div
      class="rows"
    >
      <div
        class="row error"
        data-testid="modal-error"
        id="modal-error"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          MODAL_ERROR
        </div>
      </div>
    </div>
    <div
      style="display: flex; flex-direction: row; justify-content: flex-end;"
    >
      <button
        class="sc-jTQDJr sc-iBdoyZ gmttQv bbczBv secondary"
        data-testid="cancel-modal-button"
        id="cancel-modal-button"
        style="margin: 0px 4px;"
        type="button"
      >
        Cancel
      </button>
      <button
        class="sc-jTQDJr sc-gLLuHO gmttQv lckOnr primary"
        data-testid="confirm-modal-button"
        id="confirm-modal-button"
        style="margin: 0px 4px;"
        type="submit"
      >
        Confirm
      </button>
    </div>
  </div>
</form>
`;
